File: src/opt/test/IrisTest.java
package opt.test;

import func.nn.BackpropNetworkBuilder;
import func.nn.LayeredNetwork;
import func.nn.OptNetworkBuilder;
import func.nn.backprop.BackPropagationNetwork;
import func.nn.feedfwd.FeedForwardNetwork;
import shared.*;
import shared.filt.*;
import shared.normalizer.StandardMeanAndVariance;
import shared.reader.CSVDataSetReader;
import shared.reader.DataSetReader;

import java.io.File;

/**
 * Iris nn example using network builder
 *
 * https://archive.ics.uci.edu/ml/datasets/Iris
 *
 * @author John Mansfield
 * @version 1.0
 */

public class IrisTest {

  private static int outputLayerSize;

  private static DataSet initializeData() throws Exception {
    //import data
    DataSetReader dsr = new CSVDataSetReader((new File("src/opt/test/iris.txt")).getAbsolutePath());
    DataSet ds = dsr.read();

    //standardize data
    StandardMeanAndVariance smv = new StandardMeanAndVariance();
    smv.fit(ds);
    smv.transform(ds);

    //split last attribute for label
    LabelSplitFilter lsf = new LabelSplitFilter();
    lsf.filter(ds);

    //encode label as one-hot array and get outputLayerSize
    DiscreteToBinaryFilter dbf = new DiscreteToBinaryFilter();
    dbf.filter(ds.getLabelDataSet());
    outputLayerSize=dbf.getNewAttributeCount();

    return ds;
  }

  private static LayeredNetwork runNetwork(DataSet set) {
    int percentTrain=75;

    //create backprop network using builder
    BackPropagationNetwork network = new BackpropNetworkBuilder()
      .withLayers(new int[] {25,10,outputLayerSize})
      .withDataSet(set, percentTrain)
      .withIterations(5000)
      .train();

    //create opt network using builder
    FeedForwardNetwork network2 = new OptNetworkBuilder()
      .withLayers(new int[] {25,10,outputLayerSize})
      .withDataSet(set, percentTrain)
      .withSA(100000, .975)
      .withIterations(1000)
      .train();

    return network;
  }

  public static void main(String[] args) throws Exception {
    DataSet set = initializeData();
    System.out.println(new DataSetDescription(set));
    LayeredNetwork nn = runNetwork(set);
  }
}

File: jython/iris_test.py
"""
 * Iris nn example using network builder
 *
 * https://archive.ics.uci.edu/ml/datasets/Iris
 *
 * @author John Mansfield
 * @version 1.1
"""

from __future__ import with_statement
import os
from func.nn import OptNetworkBuilder, BackpropNetworkBuilder
from shared import DataSet, Instance
from shared.filt import LabelSplitFilter, DiscreteToBinaryFilter, RandomOrderFilter, TestTrainSplitFilter
from shared.normalizer import StandardMeanAndVariance
from shared.reader import CSVDataSetReader, DataSetReader
from java.io import File

def initialize_data(input_data):
    #import data
    dsr = CSVDataSetReader((File(input_data)).getAbsolutePath())
    ds = dsr.read();
    
    #split last attribute for label
    lsf = LabelSplitFilter()
    lsf.filter(ds)

    #encode label as one-hot array and get outputLayerSize
    dbf = DiscreteToBinaryFilter()
    dbf.filter(ds.getLabelDataSet())
    output_layer_size=dbf.getNewAttributeCount()
    
    #test-train split
    percentTrain=75;
    randomOrderFilter = RandomOrderFilter();
    randomOrderFilter.filter(ds);
    testTrainSplit = TestTrainSplitFilter(percentTrain);
    testTrainSplit.filter(ds);
    train=testTrainSplit.getTrainingSet();
    test=testTrainSplit.getTestingSet();
    
    #standardize data
    smv = StandardMeanAndVariance();
    smv.fit(train);
    smv.transform(train);
    smv.transform(test);

    return train, test, output_layer_size

def run_networks(train, test, output_layer_size):
        
    #build and run backprop network using builder
    network = BackpropNetworkBuilder()\
      .withLayers([25,10,output_layer_size])\
      .withDataSet(train, test)\
      .withIterations(5000)\
      .train()
      
    #build and run RHC opt network using builder 
    network = OptNetworkBuilder()\
        .withLayers([25,10,output_layer_size])\
        .withDataSet(train, test)\
        .withRHC()\
        .withIterations(1000)\
        .train()
    
    #build and run SA opt network using builder 
    network = OptNetworkBuilder()\
        .withLayers([25,10,output_layer_size])\
        .withDataSet(train, test)\
        .withSA(15000, .95)\
        .withIterations(1000)\
        .train()
        
    #build and run GA opt network using builder 
    network = OptNetworkBuilder()\
        .withLayers([25,10,output_layer_size])\
        .withDataSet(train, test)\
        .withGA(300, 150, 50)\
        .withIterations(100)\
        .train()

def main():
    
    input_data="iris.txt"
    train, test, output_layer_size = initialize_data(input_data)
    run_networks(train, test, output_layer_size)

if __name__ == "__main__":
    main()